src/libmeasurement_kit/ooni/facebook_messenger.cpp

```diff
@@ -0,0 +1,220 @@
+                logger->info("fb_messenger: dns error for %s, %s",
+                    service.c_str(), hostname.c_str());
+            } else {
+                for (auto answer : message->answers) {
+                    if ((answer.ipv4 != "") || (answer.hostname != "")) {
+                        //XXX
+                        //std::string asn_p = options.get("geoip_asn_path", std::string{});
+                        std::string asn_p = "GeoIPASNum.dat";
+                        GeoipDatabase geoip(asn_p);
+                        ErrorOr<std::string> asn = geoip.resolve_asn(answer.ipv4);
+                        if (asn && asn.as_value() != "AS0") {
+                            logger->info("%s ipv4: %s, %s",
+                                hostname.c_str(), answer.ipv4.c_str(),
+                                asn.as_value().c_str());
+                            // if consistent, add to list for tcp connect later
+                            if (asn.as_value() == "AS32934") {
+                                (*fb_service_ips)[service].push_back(answer.ipv4);
+                            } else {
+                                (*entry)["facebook_dns_blocking"] = true;
+                            }
+                        }
+                    }
+                }
+                // if any consistent IPs, consistent = true
+                (*entry)["facebook_" + service + "_dns_consistent"] =
+                    !(*fb_service_ips)[service].empty();
+            }
+            *names_tested += 1;
+            if (names_count == *names_tested) {
+                cb(NoError(), entry, *fb_service_ips, reactor, logger);
+                return;
+            }
+        };
+    };
+
+    for (auto const& service_and_hostname : fb_service_hostnames) {
+        std::string service = service_and_hostname.first;
+        std::string hostname = service_and_hostname.second;
+        Settings options; // XXX ought this to forward from above?
+        templates::dns_query(entry, "A", "IN", hostname, "",
+                             dns_cb(service, hostname), options, reactor,
+                             logger);
+    }
+    return;
+}
+
+static void tcp_many(Error error,
+                     Var<Entry> entry,
+                     std::map<std::string, std::vector<std::string>> fb_service_ips,
+                     Var<Reactor> reactor,
+                     Var<Logger> logger,
+                     Callback<Var<Entry>> cb
+                    ) {
+    logger->info("starting tcp_many");
+    if (error) {
+        cb(entry);
+        return;
+    }
+
+    // if we find any blocked TCP later, switch this to true
+    (*entry)["facebook_dns_blocking"] = false;
+
+    int ips_count = 0;
+    for (auto const& service_and_ips : fb_service_ips) {
+        // skipping stun for now
+        if (service_and_ips.first == "stun") {
+            continue;
+        }
+        for (auto const& ip : service_and_ips.second) {
+            ips_count += 1;
```

Comment:
Is it possible to refactor this loop to count otherwise? I've turned on `-Werror` and this does not compile because the variable named `ip` is actually unused (renaming it as `_` wouldn't work).

Reply:
done